test/PostgreSQL/Test.hs

```diff
@@ -1038,18 +1066,27 @@ testInsertUniqueViolation =
       sqlErrorHint = ""}
 
 testUpsert :: SpecDb
-testUpsert =
-  describe "Upsert test" $ do
+testUpsert = focus $ describe "Upsert test" $ do
     itDb "Upsert can insert like normal" $  do
-      u1e <- EP.upsert u1 [OneUniqueName =. val "fifth"]
-      liftIO $ entityVal u1e `shouldBe` u1
+        u1e <- EP.upsert u1 [OneUniqueName =. val "fifth"]
+        liftIO $ entityVal u1e `shouldBe` u1
     itDb "Upsert performs update on collision" $  do
-      u1e <- EP.upsert u1 [OneUniqueName =. val "fifth"]
-      liftIO $ entityVal u1e `shouldBe` u1
-      u2e <- EP.upsert u2 [OneUniqueName =. val "fifth"]
-      liftIO $ entityVal u2e `shouldBe` u2
-      u3e <- EP.upsert u3 [OneUniqueName =. val "fifth"]
-      liftIO $ entityVal u3e `shouldBe` u1{oneUniqueName="fifth"}
+        u1e <- EP.upsert u1 [OneUniqueName =. val "fifth"]
+        liftIO $ entityVal u1e `shouldBe` u1
+        u2e <- EP.upsert u2 [OneUniqueName =. val "fifth"]
+        liftIO $ entityVal u2e `shouldBe` u2
+        u3e <- EP.upsert u3 [OneUniqueName =. val "fifth"]
+        liftIO $ entityVal u3e `shouldBe` u1{oneUniqueName="fifth"}
+    describe "With no updates" $ do
+        itDb "Works with no updates" $ do
+            _ <- EP.upsert u1 []
+            pure ()
+        itDb "Works with no updates, twice" $ do
+            Entity u1Key u1' <- EP.upsert u1 []
+            Entity u1Key_ u1'' <- EP.upsert u1 { oneUniqueName = "Something Else" } []
+            pure ()
```

Comment:
Unfortunately, this fails with `ErrorCall "head: empty list"`.

The reason is that:

```sql
INSERT INTO "OneUnique" (name, value)
VALUES (?, ?)
ON CONFLICT DO NOTHING
RETURNING id, name, value
```

does not return anything if we don't perform an `INSERT` - so the return is `[]` which he `head` on and it goes boom.

`DO UPDATE` always does something. But we really don't want to do something silly like `DO UPDATE SET blah = EXCLUDED.blah` because this does unnecessary writes and triggers.